### PR TITLE
Issue 5 - Move HIE Tests and convert to Tasty

### PR DESCRIPTION
**Pull request by jeffwindsor**
Tests are converted to Tasty format.  The tests compile and run from console, but most fail / timeout.
I want to check in to see if this is going the right direction, and maybe get some help wiring up the CI / Runner / Config.

### CODE REVIEW

**Comment by jeffwindsor** [comment]
helps resolve #5

**Comment by alanz** [comment]
I'm a bit embarrased to say this, but we decided to stick with hspec, because it has better rerun options.

**Comment by Bodigrim** [comment]
What's missing in `tasty-rerun`?

**Comment by alanz** [comment]
@Bodigrim possibly just my inability to look up documentation.

But I do know that if an hspec test fails, it lists a way to rerun just that test, with the appropriate command line invocation, as part of the test result. I find this invaluable.

**Comment by alanz** [comment]
And, full disclosure, I have a personal bias, as all my other OSS projects use hspec.

**Comment by Bodigrim** [comment]
`tasty-rerun` provides `--rerun` and `--rerun-all-on-success` options, pretty much similar to `hspec`. 

Full disclosure: I also have a personal bias, being a maintainer of `tasty-rerun` :) I do not mean to interfere with your project's decisions, but I'm genuinely interested what can be improved. If something is missing from `tasty-rerun`, I'll be happy to fix it.

**Comment by jeffwindsor** [comment]
closing to move to hspec

**Comment by fendor** [comment]
More discussion needed. As I look at this PR, I dont have anything to complain about on the first sight! 

@jeffwindsor Please dont undo your work, yet.

**Comment by alanz** [comment]
@Bodigrim glad to have you take an interest. The main thing I care about is if a test fails, I get something like

```
  To rerun use: --match "/Plugin/composes code actions/provides 3.8 code actions/"
```
And then I can do 

```bash
stack --stack-yaml=stack.yaml test haskell-language-server:func-test --test-arguments "--match \"/Plugin/composes code actions/provides 3.8 code actions/\""
```
To run just that test

**Comment by alanz** [comment]
One other thing that would be useful, is junit.xml output. as per https://github.com/haskell/haskell-ide-engine/blob/master/test/utils/TestUtils.hs#L306

**Comment by Bodigrim** [comment]
> One other thing that would be useful, is junit.xml output

http://hackage.haskell.org/package/tasty-ant-xml works well for me.

**Comment by alanz** [comment]
To summarize the discussion on IRC, we will continue with tasty.  For the avoidance of future doubt. Sorry to have caused so much trauma for all.

File: haskell-language-server.cabal [hunk 1]
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
...
-  -- other-modules:
-  --       Development.IDE.Test
-  --       Development.IDE.Test.Runfiles
+     ghc-options: -Werror -Wredundant-constraints
 
 library hls-test-utils
   import: agpl
   hs-source-dirs:      test/utils
-  exposed-modules:     TestUtils
+  exposed-modules:     Test.HIE.Util
```
Comment (fendor) [inline haskell-language-server.cabal:259]: This should be `Test.Hls.Util`? Or `Test.HLS.Util`, but personally, I like the former more

File: haskell-language-server.cabal [hunk 2]
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
...
+                     , Test.Tasty.Expectation
```
Comment (fendor) [inline haskell-language-server.cabal:260]: This module has been deelted.

File: haskell-language-server.cabal [hunk 3]
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
...
-  ghc-options:
-         -Wall
-         -Wredundant-constraints
-         -Wno-name-shadowing
-         -threaded -rtsopts -with-rtsopts=-N
+  import:               agpl
+  type:                 exitcode-stdio-1.0
+  default-language:     Haskell2010
+  build-tool-depends:   haskell-language-server:haskell-language-server
+                      , cabal-helper:cabal-helper-main
```
Comment (fendor) [inline haskell-language-server.cabal:213]: The whole line is in my point of view unneeded. 
cc @alanz or did you add it for a reason?

File: test/utils/Test/HIE/Util.hs [hunk 4]
```diff
@@ -1,24 +1,27 @@
 {-# LANGUAGE CPP, OverloadedStrings, NamedFieldPuns #-}
-module TestUtils
+module Test.HIE.Util
```
Comment (fendor) [inline test/utils/Test/HIE/Util.hs:2]: Out of interest, why is this renamed?

Comment (jeffwindsor) [reply test/utils/Test/HIE/Util.hs:2]: was looking for more consistent namespacing with the other testing tools, but really don't have a strong opinion.  easy for me to put but if you prefer the other way

File: haskell-language-server.cabal (hunk 3 above)
Comment (alanz) [reply haskell-language-server.cabal:213]: Well, we are running functional tests that actually run the server. If the test does not depend on the server, we are likely to end up running a functional tests against a stale build of the server.

I did not add it, but I think that is why.

Comment (fendor) [reply haskell-language-server.cabal:213]: But this `cabal-helper:cabal-helper-main` should be unnecassary, right?

Comment (fendor) [reply haskell-language-server.cabal:213]: Looks to me light an artefact from pre cabal-helper-1.0

File: test/utils/Test/HIE/Util.hs (hunk 4 above)
Comment (fendor) [reply test/utils/Test/HIE/Util.hs:2]: No, was just asking, dont have a preference either. Keep it :)

File: haskell-language-server.cabal (hunk 3 above)
Comment (alanz) [reply haskell-language-server.cabal:213]: If we are not using cabal-helper for running the server, then it can be removed

File: haskell-language-server.cabal (hunk 1 above)
Comment (jeffwindsor) [reply haskell-language-server.cabal:259]: added to next push

File: haskell-language-server.cabal (hunk 2 above)
Comment (jeffwindsor) [reply haskell-language-server.cabal:260]: done

File: haskell-language-server.cabal (hunk 3 above)
Comment (jeffwindsor) [reply haskell-language-server.cabal:213]: done

File: test/utils/Test/HIE/Util.hs (hunk 4 above)
Comment (jeffwindsor) [reply test/utils/Test/HIE/Util.hs:2]: done Test.Hls.Util

**Comment by jeffwindsor** [comment]
@fendor: 

- broken tests are marked and "pass" (https://github.com/nomeata/tasty-expected-failure/blob/master/Test/Tasty/ExpectedFailure.hs)
- added xml output (https://github.com/ocharles/tasty-ant-xml/blob/master/Test/Tasty/Runners/AntXML.hs)
- added rerun (https://github.com/ocharles/tasty-rerun/blob/master/src/Test/Tasty/Ingredients/Rerun.hs)

as of this commit I have not test the re-run or xml output successfully, but I am sure it is just a matter of getting the cli parameters to get through with the cradle

**Comment by fendor** [comment]
Awesome, I will review it in the next days in depth!

**Comment by Bodigrim** [comment]
> @Bodigrim glad to have you take an interest. The main thing I care about is if a test fails, I get something like
> 
> ```
>   To rerun use: --match "/Plugin/composes code actions/provides 3.8 code actions/"
> ```

Your wish is [granted](https://hackage.haskell.org/package/tasty-1.4.2/changelog) :)
